### PR TITLE
add some allocate cases

### PR DESCRIPTION
**Pull request by alcorf-mizar**
Signed-off-by: alcorf-mizar <yihang.von@gmail.com>

### CODE REVIEW

File: test/e2e/allocate.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
+package e2e
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:1]: Add copy right

File: test/e2e/allocate.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+import (
+	"context"
+	. "github.com/onsi/ginkgo"
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:5]: nit: add a blank line before

File: test/e2e/allocate.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/api/resource"
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:7]: move down and well group all `k8s.io` prefix lib

File: test/e2e/util.go [hunk 4]
```diff
@@ -1379,3 +1468,39 @@ func isPodScheduled(pod *v1.Pod) bool {
...
+	})
+
+	if len(pods.Items) == 1 {
+		return pods.Items[0].Spec.NodeName, nil
+	}
+
+	return "", fmt.Errorf("get pods num not equal 1")
+}
+
+func compareNodename(ctx *testContext, jobA, jobB *batchv1alpha1.Job) (err error) {
```
Comment (hzxuzhonghu) [inline test/e2e/util.go:1497]: We can not guarantee the given job is of one replica.  May cause misuse by other developers

File: test/e2e/allocate.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+					req: slot1,
+					min: 1,
+					rep: 1,
+				},
+			},
+		}
+
+		job.name = "j1-n3"
+		node3Job := createJobWithCondition(ctx, job, "", map[string]string{
+			corev1.LabelHostname: "node3",
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:218]: how do you get "node3"

File: test/e2e/allocate.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+		Expect(err).NotTo(HaveOccurred())
+
+		err = compareNodename(ctx, lowReqJob, highReqJob)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("schedule job when resource is NOT enough", func() {
+		ctx := initTestContext(options{
+			nodesNumLimit: 2,
+			nodesResourceLimit: corev1.ResourceList{
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:60]: does this mean the resource specified is idle?

File: test/e2e/allocate.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+		err = waitJobReady(ctx, highReqJob)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = compareNodename(ctx, lowReqJob, highReqJob)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("schedule job when resource is NOT enough", func() {
+		ctx := initTestContext(options{
+			nodesNumLimit: 2,
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:59]: what does this mean?

File: test/e2e/allocate.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+		pg := &schedulingv1beta1.PodGroup{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: ctx.namespace,
+				Name:      pgName,
+			},
+			Spec: schedulingv1beta1.PodGroupSpec{
+				MinMember:    1,
+				MinResources: &thirtyCPU,
+			},
+			Status: schedulingv1beta1.PodGroupStatus{
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:127]: remove, it is common sense to create without status specified.

File: test/e2e/allocate.go [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+		err = waitJobReady(ctx, existJob)
+		Expect(err).NotTo(HaveOccurred())
+
+		job.name = "allocate"
+		allocateJob := createJobWithCondition(ctx, job, pgName, nil)
+		err = waitJobPending(ctx, allocateJob)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("allocate don't work when queue NOT exist", func() {
```
Comment (hzxuzhonghu) [inline test/e2e/allocate.go:161]: I donot think this case will test anything on scheduling, the job creation is denied by admission.
So i would suggest removing it

File: test/e2e/allocate.go (hunk 5 above)
Comment (hzxuzhonghu) [reply test/e2e/allocate.go:218]: Do you mean a non-exist node?

File: test/e2e/util.go [hunk 10]
```diff
@@ -333,6 +335,93 @@ func createJob(context *testContext, jobSpec *jobSpec) *batchv1alpha1.Job {
 	return job
 }
 
+func createJobWithCondition(ctx *testContext, jobSpec *jobSpec, pgName string, nodeSelector map[string]string) *batchv1alpha1.Job {
```
Comment (hzxuzhonghu) [inline test/e2e/util.go:338]: It is not clear what condition is from the function name

File: test/e2e/allocate.go (hunk 9 above)
Comment (alcorf-mizar) [reply test/e2e/allocate.go:161]: I agree with you. I'll remove it.

File: test/e2e/allocate.go (hunk 8 above)
Comment (alcorf-mizar) [reply test/e2e/allocate.go:127]: Ok

File: test/e2e/allocate.go (hunk 5 above)
Comment (alcorf-mizar) [reply test/e2e/allocate.go:218]: I will use other conditions to test allocate when job not satisify predicates.

File: test/e2e/allocate.go (hunk 6 above)
Comment (alcorf-mizar) [reply test/e2e/allocate.go:60]: This means the resources specified is required(Prepare form cluster nodes).

**Pull request merged**